File: Android/app/src/androidTest/java/PageObjects/LoginPageObjects/LoginPageObject.kt
/*
 * Copyright (c) 2017-present, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package pageobjects.loginpageobjects

import android.util.Log
import android.view.accessibility.AccessibilityWindowInfo
import androidx.test.uiautomator.UiObject
import androidx.test.uiautomator.UiSelector
import androidx.test.platform.app.InstrumentationRegistry
import pageobjects.BasePageObject

const val USERNAME_RESOURCE_ID = "username"
const val PASSWORD_RESOURCE_ID = "password"
const val LOGIN_RESOURCE_ID = "Login"

/**
 * Chrome Custom Tab close button. Under forced advanced authentication the login form is rendered
 * in a Chrome Custom Tab, so its presence is the marker every tab-facing helper keys off.
 */
const val CHROME_CLOSE_BUTTON_ID = "com.android.chrome:id/close_button"

/**
 * Short timeout for checking optional local Chrome UI elements that either appear immediately or
 * not at all (e.g. first-run dialogs). These are not dependent on server-side rendering.
 */
private const val QUICK_CHECK_TIMEOUT: Long = 500
private const val KEYBOARD_DISMISS_TIMEOUT: Long = 2_000

private const val LOCAL_NETWORK_PERMISSION_MESSAGE =
    "wants to access other devices on your local network"
private const val CHROME_PERMISSION_MESSAGE_ID = "com.android.chrome:id/text"
private const val CHROME_NEGATIVE_BUTTON_ID = "com.android.chrome:id/negative_button"

/**
 * Ceiling for clearing Chrome's First Run Experience, which on a cold profile is a slow multi-page
 * sequence. Longer than [timeout]; only fully spent when no tab ever appears.
 */
private const val FRE_DISMISS_TIMEOUT: Long = 30_000

/**
 * Created by bpage on 2/21/18.
 *
 * Drives the Salesforce login form. Under forced advanced authentication the form is rendered in a
 * Chrome Custom Tab, which runs in a separate process (com.android.chrome) that Espresso cannot
 * reach; UiAutomator cross-process lookups are used instead. Fields are matched by their web
 * resource id first, then fall back to the generic EditText/Button classes when Chrome does not
 * surface the ids (QUERY_ALL_PACKAGES in the androidTest manifest grants the cross-process access).
 *
 * LEGACY UPGRADE AUTOMATION (SDK 12.x AND 13.x): [expectAdvancedAuthentication] is false while
 * logging into v12.2.0 and v13.2.1 apps, whose login form is still in the in-app WebView. Remove
 * the parameter and conditional branch after both 12.x and 13.x upgrade coverage are retired.
 */
class LoginPageObject(
    private val expectAdvancedAuthentication: Boolean = true,
) : BasePageObject() {

    fun setUsername(name: String) {
        prepareLoginSurface()
        Log.i("uia", "Waiting for username field to be present.")
        val usernameField = waitForLoginControl(
            UiSelector().resourceId(USERNAME_RESOURCE_ID),
            UiSelector().className(editTextClass).instance(0),
            "Username field not found."
        )
        usernameField.click()
        usernameField.setText(name)
    }

    fun setPassword(password: String) {
        prepareLoginSurface()
        Log.i("uia", "Waiting for password field to be present.")
        val passwordField = waitForLoginControl(
            UiSelector().resourceId(PASSWORD_RESOURCE_ID),
            UiSelector().className(editTextClass).instance(0),
            "Password field not found."
        )
        passwordField.click()
        passwordField.setText(password)
    }

    fun tapLogin() {
        dismissKeyboardIfPresent()
        val loginButton = waitForLoginControl(
            UiSelector().resourceId(LOGIN_RESOURCE_ID),
            UiSelector().className("android.widget.Button").textContains("Log In"),
            "Log In button not found.",
            timeout
        )
        loginButton.click()
    }

    private fun dismissKeyboardIfPresent() {
        val uiAutomation = InstrumentationRegistry.getInstrumentation().uiAutomation
        if (uiAutomation.windows.none { it.type == AccessibilityWindowInfo.TYPE_INPUT_METHOD }) {
            return
        }

        device.pressBack()
        val deadline = System.currentTimeMillis() + KEYBOARD_DISMISS_TIMEOUT
        while (System.currentTimeMillis() < deadline &&
            uiAutomation.windows.any { it.type == AccessibilityWindowInfo.TYPE_INPUT_METHOD }) {
            device.waitForIdle(QUICK_CHECK_TIMEOUT)
        }
        if (uiAutomation.windows.any { it.type == AccessibilityWindowInfo.TYPE_INPUT_METHOD }) {
            throw AssertionError("Keyboard did not close before tapping Log In.")
        }
    }

    /**
     * Advanced authentication opens the login form in Chrome, whose first-run UI may need clearing.
     * LEGACY UPGRADE AUTOMATION (SDK 12.x AND 13.x): v12.2.0 and v13.2.1 keep the login form in the
     * app's WebView, so their initial-login phase skips Chrome preparation. Remove this branch only
     * after both 12.x and 13.x upgrade coverage are retired.
     */
    private fun prepareLoginSurface() {
        if (expectAdvancedAuthentication) {
            skipGoogleSignIn()
        }
    }

    private fun waitForLoginControl(
        primarySelector: UiSelector,
        fallbackSelector: UiSelector,
        errorMessage: String,
        waitTimeout: Long = timeout * 5
    ): UiObject {
        val controls = listOf(primarySelector, fallbackSelector).map(device::findObject)
        val deadline = System.currentTimeMillis() + waitTimeout
        while (System.currentTimeMillis() < deadline) {
            blockLocalNetworkAccessIfPresent()
            for (control in controls) {
                if (control.waitForExists(QUICK_CHECK_TIMEOUT)) {
                    blockLocalNetworkAccessIfPresent()
                    return control
                }
            }
        }
        throw AssertionError(errorMessage)
    }

    /**
     * Blocks Chrome's optional local-network request, which is unrelated to remote OAuth login.
     * The message check prevents Chrome's shared negative-button id from dismissing other dialogs.
     */
    private fun blockLocalNetworkAccessIfPresent() {
        val permissionMessage = device.findObject(
            UiSelector()
                .resourceId(CHROME_PERMISSION_MESSAGE_ID)
                .textContains(LOCAL_NETWORK_PERMISSION_MESSAGE)
        )
        if (!permissionMessage.exists()) {
            return
        }

        Log.i("uia", "Blocking Chrome local-network access prompt.")
        val blockButton = device.findObject(UiSelector().resourceId(CHROME_NEGATIVE_BUTTON_ID))
        if (!blockButton.waitForExists(QUICK_CHECK_TIMEOUT)) {
            throw AssertionError("Chrome local-network Block button not found.")
        }
        blockButton.click()
        if (!permissionMessage.waitUntilGone(timeout)) {
            throw AssertionError("Chrome local-network access prompt did not close.")
        }
    }

    /**
     * Clears Chrome's First Run Experience so the Custom Tab can render. On a cold Chrome the tab's
     * first launch is covered by a multi-page FRE that renders slowly, hiding the tab contents every
     * tab-facing helper keys off. Loops until the tab is in front, dismissing whichever FRE control
     * shows each pass. Idempotent: a warm Chrome already showing the tab returns at once.
     */
    fun skipGoogleSignIn() {
        val deadline = System.currentTimeMillis() + FRE_DISMISS_TIMEOUT
        while (System.currentTimeMillis() < deadline) {
            if (isCustomTabDisplayed()) {
                return
            }
            if (!dismissOneFreControl()) {
                // Nothing to dismiss yet; the next FRE page may still be rendering, so re-check.
                device.waitForIdle(QUICK_CHECK_TIMEOUT)
            }
        }
        throw AssertionError("Chrome Custom Tab never appeared after ${FRE_DISMISS_TIMEOUT}ms.")
    }

    /**
     * Dismisses a single FRE control if one is on screen, returning true if it clicked something.
     * Decline buttons are tried before accept buttons so the flow advances without a Google sign-in;
     * each is matched by resource id first, then by its en-locale label.
     */
    private fun dismissOneFreControl(): Boolean {
        val dismissByIdOrText = listOf(
            "com.android.chrome:id/signin_fre_dismiss_button",
            CHROME_NEGATIVE_BUTTON_ID,
        )
        for (resourceId in dismissByIdOrText) {
            val button = device.findObject(UiSelector().resourceId(resourceId))
            if (button.waitForExists(QUICK_CHECK_TIMEOUT)) {
                button.click()
                return true
            }
        }
        for (label in listOf("Use without an account", "No thanks", "No Thanks")) {
            val button = device.findObject(UiSelector().textContains(label))
            if (button.exists()) {
                button.click()
                return true
            }
        }

        // The initial UMA/ToS page has no decline option; accepting it is required to advance.
        val acceptByIdOrText = listOf(
            "com.android.chrome:id/terms_accept",
            "com.android.chrome:id/positive_button",
        )
        for (resourceId in acceptByIdOrText) {
            val button = device.findObject(UiSelector().resourceId(resourceId))
            if (button.waitForExists(QUICK_CHECK_TIMEOUT)) {
                button.click()
                return true
            }
        }
        for (label in listOf("Accept & continue", "Got it")) {
            val button = device.findObject(UiSelector().textContains(label))
            if (button.exists()) {
                button.click()
                return true
            }
        }

        return false
    }

    /**
     * True when a Chrome Custom Tab is currently in front, detected via its close button.
     * Used to confirm the advanced-auth login form is being driven in the tab rather than in-app.
     */
    fun isCustomTabDisplayed(): Boolean {
        val closeButton = device.findObject(
            UiSelector().resourceId(CHROME_CLOSE_BUTTON_ID)
        )
        return closeButton.waitForExists(QUICK_CHECK_TIMEOUT)
    }
}

File: iOS/PageObjects/LoginPageObjects/LoginPageObject.swift
/*
 * Copyright (c) 2017-present, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
//
//  LoginPageObject.swift
//  MobileSDKUITest
//
//  Created by Brandon Page on 2/21/18.
//

import Foundation
import XCTest

class LoginPageObject {
    let app:XCUIApplication
    let timeout:double_t = 10
    private let expectAdvancedAuthentication: Bool

    init(testApp: XCUIApplication, expectAdvancedAuthentication: Bool = true) {
        app = testApp
        self.expectAdvancedAuthentication = expectAdvancedAuthentication
    }

    /// The web content that hosts the login form. Pre-14.0 upgrade tests use the app's top-level
    /// WebView. Under forced advanced auth the form is served in an external browser
    /// (ASWebAuthenticationSession). Some app types (notably hybrid) trigger authentication more
    /// than once on cold launch, which stacks multiple browser sheets — each an identical
    /// `webViews.webViews.webViews` subtree with its own username/password fields. Drive the
    /// frontmost (last-presented, top-of-stack) browser and scope every field/button query to it.
    private var loginWebView: XCUIElement {
        // LEGACY UPGRADE AUTOMATION (SDK 12.x AND 13.x): v12.2.0 and v13.2.1 use their
        // top-level WebView. Remove this branch after both 12.x and 13.x coverage are retired.
        if !expectAdvancedAuthentication {
            return app.webViews.firstMatch
        }
        let webViews = app.webViews.webViews.webViews
        let count = webViews.count
        return count > 1 ? webViews.element(boundBy: count - 1) : webViews.firstMatch
    }

    func setUsername(name: String) -> Void {
        // Wait for the configured login surface to finish rendering before touching fields.
        waitForLoginFormReady()
        hideKeyboard()
        let nameField = loginWebView.textFields.firstMatch
        _ = nameField.waitForExistence(timeout: timeout * 12)
        // The browser pre-populates the username from the OAuth login_hint. A pre-filled web field
        // does not reliably attach the software keyboard on tap, so typing into it fails with
        // "Neither element nor any descendant has keyboard focus". When the field already holds a
        // value, trust the prefill and skip typing — the caller will advance with the "Log In"
        // button. Only type when the field is genuinely empty.
        let existing = nameField.value as? String ?? ""
        if !existing.isEmpty && existing != "Username" {
            return
        }
        setFieldValue(nameField, value: name)
    }

    func setPassword(password: String) -> Void {
        hideKeyboard()
        // Tapping "Log In" on the username step navigates the browser to the password page; allow the
        // same generous browser-page-load window used for the initial form before touching the field.
        let passwordField = loginWebView.secureTextFields.firstMatch
        _ = passwordField.waitForExistence(timeout: timeout * 12)
        setFieldValue(passwordField, value: password)
    }

    /// Enters `value` into a browser login-form field. The external browser may pre-populate a
    /// field (e.g. a remembered username), and tapping an already-filled web field does not always
    /// attach the software keyboard — typing then fails with "no keyboard focus". So tap to focus,
    /// skip typing when the field already holds the desired value, and only type otherwise. Mirrors
    /// the SDK AuthFlowTester UITests' setTextField for the forced-advanced-auth browser flow.
    private func setFieldValue(_ field: XCUIElement, value: String) {
        field.tap()
        sleep(1)
        if (field.value as? String) == value {
            return
        }
        field.typeText(value)
    }

    /// Waits for the configured login form to be fully loaded and interactive. Browser-based
    /// advanced authentication and the legacy in-app WebView both render the fields asynchronously.
    func waitForLoginFormReady() -> Void {
        let webViewTextField = loginWebView.textFields.firstMatch
        _ = webViewTextField.waitForExistence(timeout: timeout * 12)
    }

    /// Submits the current login step. Under forced advanced authentication the login form is served
    /// in the external browser (ASWebAuthenticationSession), which renders an on-page "Log In" button.
    /// Tapping that button is how a user advances, and it does not require keyboard focus — synthesizing
    /// a Return keypress into a web field fails ("Neither element nor any descendant has keyboard
    /// focus") whenever the field was pre-filled and never attached the software keyboard. Falls back
    /// to a Return keypress only if no button is present. Callers drive the form the same way a user
    /// would: (accept prefilled or type) username, submit; type password, submit.
    func tapLogin() -> Void {
        hideKeyboard()
        let webView = loginWebView
        // Salesforce's identity-first browser form labels the button "Log In"; match loosely to also
        // catch "Log In to Sandbox" and locale/label variants.
        let loginButton = webView.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'Log In'")).firstMatch
        if loginButton.waitForExistence(timeout: timeout) && loginButton.isHittable {
            loginButton.tap()
            return
        }
        // Fallback: submit the on-screen field with Return (only works if the keyboard is attached).
        let secureField = webView.secureTextFields.firstMatch
        let field = secureField.exists ? secureField : webView.textFields.firstMatch
        field.typeText(XCUIKeyboardKey.return.rawValue)
    }

    /// Dismisses the external browser (ASWebAuthenticationSession) without completing login. The
    /// browser chrome exposes a "Close" button on current iOS versions and "Cancel" on earlier ones.
    func tapBack() -> Void {
        let topBar = app.otherElements["TopBrowserBar"]
        var closeButton = topBar.buttons["Close"]
        if !closeButton.waitForExistence(timeout: timeout) {
            // Earlier iOS versions use "Cancel"; only that fallback needs its own wait.
            closeButton = topBar.buttons["Cancel"]
            _ = closeButton.waitForExistence(timeout: timeout)
        }
        closeButton.tap()
    }
    
    func hideKeyboard() -> Void {
        let continueButton = app.otherElements["UIContinuousPathIntroductionView"]
        if continueButton.exists {
            continueButton.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.5)).tap()
        }
        
        let doneButton = app.toolbars.matching(identifier: "Toolbar").buttons["Done"]
        if doneButton.exists {
            doneButton.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.5)).tap()
        }
    }
}
